labview-interop/src/errors.rs

```diff
@@ -1,89 +1,650 @@
+    }
+}
+
+/// LVError is a generic Labview Error
+/// that can retrieve a error description if the link feature is enabled
+#[derive(Debug, Clone, Copy, Eq, PartialEq, Error)]
+pub struct LVError {
+    code: LVStatusCode,
+}
+
+impl ToLvError for LVError {
+    fn source(&self) -> std::borrow::Cow<'_, str> {
+        "Rust".into()
+    }
+
+    fn code(&self) -> LVStatusCode {
+        self.code
+    }
+
+    #[cfg(feature = "link")]
+    fn description(&self) -> Cow<'static, str> {
+        static DEFAULT_STRING: &str = "LabVIEW-Interop: Description not retrievable";
+        let mut error_text_ptr = MaybeUninit::<LStrHandle>::uninit();
+
+        let memory_api = match labview::memory_api() {
+            Ok(api) => api,
+            Err(_) => return Cow::Borrowed(DEFAULT_STRING),
+        };
+
+        unsafe {
+            if memory_api
+                .error_code_description(self.code.0, error_text_ptr.as_mut_ptr() as *mut usize)
+            {
+                let error_text_ptr = error_text_ptr.assume_init();
+                let desc = error_text_ptr.to_rust_string().to_string();
+                return Cow::Owned(desc);
+            }
+        }
+        Cow::Borrowed(DEFAULT_STRING)
+    }
+}
+
+impl From<LVStatusCode> for core::result::Result<(), LVError> {
+    fn from(code: LVStatusCode) -> Self {
+        match code {
+            LVStatusCode::SUCCESS => Ok(()),
+            _ => Err(LVError { code }),
+        }
+    }
+}
+
+impl Display for LVError {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(
+            f,
+            "{}: {}",
+            self.code,
+            <LVError as ToLvError>::description(self)
+        )
+    }
+}
+
+/// MgError is the subset of LabVIEW errors that may occur when dealing with the memory manager
+/// So in the context of Rust-LabVIEW-interop these are the kind of labview errors we may trigger within the library
+///
+///
+/// The `MgError` / `MgErrorCode` implement From in both directions. Additionally IntoPrimitive and TryFromPrimitive is derived
+/// to enable the conversion from and to int primitives.
+///
+/// LabVIEW official general error list:
+/// https://www.ni.com/docs/en-US/bundle/labview-api-ref/page/errors/general-labview-error-codes.html
+/// more complete inofficial lists:
+/// https://labviewwiki.org/wiki/LabVIEW_Error_Code_Family   /  https://labviewwiki.org/wiki/List_of_errors
+///
+/// Accoring to https://www.ni.com/docs/en-US/bundle/labview/page/labview-manager-function-errors.html
+/// the memory manager only uses a subset of this huge error list. The subset is implemented in `MgError` using
+/// the official abbreviations.
+
+/// in https://www.ni.com/docs/en-US/bundle/labview/page/labview-manager-function-errors.html
+/// `MgError` implements Error on top of the `MgErrorCode` and includes a description
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Error, IntoPrimitive, TryFromPrimitive)]
+#[repr(i32)]
+pub enum MgError {
```

Comment:
I might be tempted to split this out into submodules because of the size. But happy to do that after merging this request.